app/src/swap.c

```diff
@@ -0,0 +1,249 @@
+    if (ctx == NULL) {
+        return err;
+    }
+    // Check method.
+    const char * valid_tx_method = "Balances Transfer";
+    char tmp_str[80] = {0};
+    snprintf(tmp_str, sizeof(tmp_str), "%s %s", _getMethod_ModuleName(ctx->tx_obj->transactionVersion,
+                                                                    ctx->tx_obj->callIndex.moduleIdx),
+                                                _getMethod_Name(ctx->tx_obj->transactionVersion,
+                                                                ctx->tx_obj->callIndex.moduleIdx,
+                                                                ctx->tx_obj->callIndex.idx));
+
+    if (strncmp(tmp_str, &valid_tx_method[0], strlen(valid_tx_method)) != 0) {
+         PRINTF("Wrong swap tx method (%s, should be : %s).\n",tmp_str,valid_tx_method);
+         ZEMU_LOGF(100, "Wrong swap tx method (%s, should be : %s).\n",tmp_str,valid_tx_method)
+         return parser_swap_tx_wrong_method;
+    }
+
+    // Check transaction method arguments number. Balance transfer Should be 2 (for tx v18).
+    if (_getMethod_NumItems(ctx->tx_obj->transactionVersion,
+                           ctx->tx_obj->callIndex.moduleIdx,
+                           ctx->tx_obj->callIndex.idx) != 2) {
+        ZEMU_LOGF(50, "Wrong swap tx method arguments count.\n")
+        return parser_swap_tx_wrong_method_args_num;
+    }
+
+    // Check destination address.
+    MEMZERO(tmp_str,sizeof(tmp_str));
+    uint8_t pageCount = 0;
+    err = _getMethod_ItemValue(ctx->tx_obj->transactionVersion,
+                            &ctx->tx_obj->method,
+                            ctx->tx_obj->callIndex.moduleIdx, ctx->tx_obj->callIndex.idx, 0,
+                            tmp_str, sizeof(tmp_str),
+                            0, &pageCount);
+
+    if (err != parser_ok) {
+        ZEMU_LOGF(50, "Could not parse swap tx destination address.");
+        return err;
+    }
+
+    if (strncmp(tmp_str, &(G_swap_state.destination_address[0]), sizeof(G_swap_state.destination_address)) != 0) {
+         ZEMU_LOGF(100, "Wrong swap tx destination address (%s, should be : %s).\n", tmp_str, G_swap_state.destination_address)
+         return parser_swap_tx_wrong_dest_addr;
+    }
+
+    // Check amount.
+    MEMZERO(tmp_str, sizeof(tmp_str));
+    err = _getMethod_ItemValue(ctx->tx_obj->transactionVersion,
+                            &ctx->tx_obj->method,
+                            ctx->tx_obj->callIndex.moduleIdx, ctx->tx_obj->callIndex.idx, 1,
+                            tmp_str, sizeof(tmp_str),
+                            0, &pageCount);
+
+    if(err != parser_ok)
+    {
+        ZEMU_LOGF(50, "Could not parse swap tx amount.");
+        return err;
+    }
+
+    char tmp_amount[50] = {0};
+    pd_CompactBalance_t amount = {.value.ptr = G_swap_state.amount, .value.len = G_swap_state.amount_length};
+    err = _toStringCompactBalance(&amount, tmp_amount, sizeof(tmp_amount), 0, &pageCount);
```

Comment:
I think the way you store the amount in `G_swap_state.amount` is wrong. 

When I run the tests (after fixing the amount_length assignment) I get `DOT 0.374145024` instead of `DOT 1.234567` in `tmp_amount` after `_toStringCompactBalance` call.

Reply:
Let's check if was that big endian problem 👌🏻

Reply:
@carlosala the problem you have is that you are using `_toStringCompactBalance` on `G_swap_state.amount` that is not a compact integer. It is a standard uint8 array of length 16 (equivalent to a uint of 128 bits).

In my swap PR I have used the special function `bytes_amount_to_print_str`  to convert this array to a printable amount (cf. https://github.com/LedgerHQ/app-polkadot/blob/e5c924ae53742814f241f97410b100d2fd3b1160/app/src/swap/swap_lib_calls.c#L62)

You need something similar.